source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -12,70 +12,150 @@ namespace Extensions {
+      parameter_mutations_.emplace_back(std::make_unique<ParameterMutationAppend>(
+          mutation.append().record().key(), std::move(value_or_error.value()),
+          mutation.append().action()));
+
+    } else if (!mutation.remove().empty()) {
+      parameter_mutations_.emplace_back(
+          std::make_unique<ParameterMutationRemove>(mutation.remove()));
+    } else {
+      creation_status = absl::InvalidArgumentError("No mutation specified.");
+      return;
+    }
+  }
+}
+
+void Mutations::mutateRequestHeaders(Http::RequestHeaderMap& headers,
+                                     const Formatter::HttpFormatterContext& context,
                                      const StreamInfo::StreamInfo& stream_info) const {
-  request_mutations_->evaluateHeaders(headers, ctx, stream_info);
+  request_mutations_->evaluateHeaders(headers, context, stream_info);
+
+  if (parameter_mutations_.empty() || headers.Path() == nullptr) {
+    return;
+  }
+
+  // Mutate query parameters.
+  Http::Utility::QueryParamsMulti params =
+      Http::Utility::QueryParamsMulti::parseAndDecodeQueryString(headers.getPathValue());
+  for (const auto& mutation : parameter_mutations_) {
+    mutation->mutateQueryParameter(params, context, stream_info);
+  }
+  headers.setPath(params.replaceQueryString(headers.Path()->value()));
 }
 
-void Mutations::mutateResponseHeaders(Http::HeaderMap& headers,
-                                      const Formatter::HttpFormatterContext& ctx,
+void Mutations::mutateResponseHeaders(Http::ResponseHeaderMap& headers,
+                                      const Formatter::HttpFormatterContext& context,
                                       const StreamInfo::StreamInfo& stream_info) const {
-  response_mutations_->evaluateHeaders(headers, ctx, stream_info);
+  response_mutations_->evaluateHeaders(headers, context, stream_info);
 }
 
-PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config)
-    : mutations_(config.mutations()) {}
+PerRouteHeaderMutation::PerRouteHeaderMutation(const PerRouteProtoConfig& config,
+                                               absl::Status& creation_status)
+    : mutations_(config.mutations(), creation_status) {}
 
-HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config)
-    : mutations_(config.mutations()),
+HeaderMutationConfig::HeaderMutationConfig(const ProtoConfig& config, absl::Status& creation_status)
+    : mutations_(config.mutations(), creation_status),
       most_specific_header_mutations_wins_(config.most_specific_header_mutations_wins()) {}
 
-Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
-  Formatter::HttpFormatterContext ctx{&headers};
-  config_->mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
-
+void HeaderMutation::initializeRouteConfigs(Http::StreamFilterCallbacks* callbacks) {
+  route_configs_initialized_ = true;
```

Comment:
> I'm not sure if the boolean flag is really needed. Can `route_configs_` be used instead?

**One of core target here is to ensure the whole filter will use same config in both decoding phase and encoding phase. (If the filter is called in both phase.)**
But `route_configs_`'s empty check cannot achieved this target.

For example, the request is initially matched to route A. At the decoding phase, the request gets a empty `route_configs_` because the route A has no header mutation configs. Then another filter updated the request headers and refreshed route and the request is matched to another route B. The route B may have some header mutation configs. At the encoding phase, the header mutation may try to initialize the `route_configs_` again.

Reply:
Can this comment be added to the beginning of this function?
I think it will improve the life of future reviewers and simplify code changes here.

Reply:
Got it.